mlbench/refimpls/pytorch/optim/optimizer.py

```diff
@@ -15,16 +18,174 @@ def get_optimizer(options, model):
+        sparse_grad_size (int): Size of the sparsified gradients vector.
+
+    """
+
+    def __init__(self, params, lr=required, weight_decay=0, sparse_grad_size=10):
+
+        if lr is not required and lr < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if weight_decay < 0.0:
+            raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+
+        defaults = dict(lr=lr, weight_decay=weight_decay)
+
+        super(sparsified_SGD, self).__init__(params, defaults)
+
+        self.__create_gradients_memory()
+        self.__create_weighted_average_params()
+
+        self.num_coordinates = sparse_grad_size
+
+    def __setstate__(self, state):
+        super(sparsified_SGD, self).__setstate__(state)
+
+    def __create_weighted_average_params(self):
+        """ Create a memory to keep the weighted average of parameters in each iteration """
+        for group in self.param_groups:
+            for p in group['params']:
+                param_state = self.state[p]
+                param_state['estimated_w'] = torch.zeros_like(p.data)
+                p.data.normal_(0, 0.01)
+                param_state['estimated_w'].copy_(p.data)
+
+    def __create_gradients_memory(self):
+        """ Create a memory to keep gradients that are not used in each iteration """
+        for group in self.param_groups:
+            for p in group['params']:
+                param_state = self.state[p]
+                param_state['memory'] = torch.zeros_like(p.data)
+
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+
+            weight_decay = group['weight_decay']
+
+            for p in group['params']:
+
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
+
+                if weight_decay != 0:
+                    d_p.add_(weight_decay, p.data)
+                p.data.add_(-d_p)
+
+        return loss
+
+    def sparsify_gradients(self, model, lr, random_sparse):
+        """ Calls one of the sparsification functions (random or blockwise)"""
+        if random_sparse:
+            return self._random_sparsify(model, lr)
+        else:
+            return self._block_sparsify(model, lr)
+
+    def _random_sparsify(self, model, lr):
+        """
+        Sparsify the gradients vector by selecting 'k' of them randomly.
+        param model: learning model
+        param lr: learning rate
+        return: sparsified gradients vector ('k' gradients and their indices)
+        """
+        params_sparse_tensors = []
+
+        for ind, param in enumerate(model.parameters()):
+            self.state[param]['memory'] += param.grad.data * lr[0]
```

Comment:
If a layer or similar gets manually assigned a different tensor, the dict-key here would change (it's `id(parameter)` which maps to the python object identifier). This could be prevented by using `model.named_parameters()` which returns a unique name for each parameter along with the values. Then that name could be used as dict key.

Reply:
I cannot use the name as the dict key, because the key here is not the name of the parameter, it's the parameter itself.